File: tools/trainers/pi_kuhul_engine.py
"""
π-KUHUL: Mayan Math to Python Bridge
Complete field-theoretic inference engine with Maya calendar, vigesimal arithmetic, and π-phase geometry.

THIS IS A RAG ENGINE:
  Corpus    = Token states in 100D manifold (the "documents")
  Retrieval = Geodesic attention (interference-weighted neighbour lookup)
  Augment   = Shard transition based on phase coherence
  Generate  = Field evolution step

Haab months = K'UHUL phases (not coincidence — the calendar IS the execution schedule):
  Pop   Wo   Sip  Sotz  Sek  Xul  Yaxk'in  Mol  Ch'en  Yax...
  Pop → Wo → Sek → Ch'en → Xul

52 cards = SMGM-16 CardSlots = Calendar Round (52 Haab years)
Token.phase = OpticalNode.sh[] (SH wave state)
shard_id = optical cluster membership
interference() = arccos(q·k^T) geodesic attention
coherence_history = EntropyField values
"""

import math
import numpy as np
from typing import List, Dict, Tuple, Optional, Union
from dataclasses import dataclass
from datetime import datetime, timedelta

# ============================================================
# MAYAN DIGIT SYMBOLS TO NUMERIC VALUES
# ============================================================

MAYAN_DIGITS = {
    '•': 1, '••': 2, '•••': 3, '••••': 4,
    '⎯': 5, '⎯⎯': 10, '⎯⎯⎯': 15,
    '•⎯': 6, '••⎯': 7, '•••⎯': 8, '••••⎯': 9,
    '•⎯⎯': 11, '••⎯⎯': 12, '•••⎯⎯': 13, '••••⎯⎯': 14,
    '•⎯⎯⎯': 16, '••⎯⎯⎯': 17, '•••⎯⎯⎯': 18, '••••⎯⎯⎯': 19,
    '𑂐': 0, '🐚': 0, '○': 0, '·': 0
}

REVERSE_MAYAN = {
    0:'𑂐', 1:'•', 2:'••', 3:'•••', 4:'••••', 5:'⎯',
    6:'•⎯', 7:'••⎯', 8:'•••⎯', 9:'••••⎯', 10:'⎯⎯',
    11:'•⎯⎯', 12:'••⎯⎯', 13:'•••⎯⎯', 14:'••••⎯⎯', 15:'⎯⎯⎯',
    16:'•⎯⎯⎯', 17:'••⎯⎯⎯', 18:'•••⎯⎯⎯', 19:'••••⎯⎯⎯'
}

def mayan_to_number(glyph: str) -> int:
    if glyph in MAYAN_DIGITS:
        return MAYAN_DIGITS[glyph]
    if all(c in '•⎯𑂐' for c in glyph):
        value = 0
        for char in glyph:
            value = value * 20 + MAYAN_DIGITS.get(char, 0)
        return value
    raise ValueError(f"Invalid Mayan glyph: {glyph}")

def number_to_mayan(n: int) -> str:
    if not isinstance(n, int) or n < 0: return '𑂐'
    if n == 0: return '𑂐'
    digits = []
    while n > 0:
        digits.insert(0, n % 20)
        n //= 20
    return ''.join(REVERSE_MAYAN[d] for d in digits)

# ============================================================
# MAYAN LONG COUNT CALENDAR
# ============================================================

@dataclass
class LongCount:
    baktun: int = 13; katun: int = 0; tun: int = 0; uinal: int = 0; kin: int = 0

    def to_days(self) -> int:
        return (self.baktun*144000 + self.katun*7200 +
                self.tun*360 + self.uinal*20 + self.kin)

    @classmethod
    def from_days(cls, days: int) -> 'LongCount':
        b=days//144000; days%=144000; k=days//7200; days%=7200
        t=days//360; days%=360; u=days//20; kin=days%20
        return cls(b,k,t,u,kin)

    def advance(self, days: int=1) -> 'LongCount':
        return LongCount.from_days(self.to_days()+days)

    def __str__(self): return f"{self.baktun}.{self.katun}.{self.tun}.{self.uinal}.{self.kin}"

@dataclass
class Tzolkin:
    number: int; day_sign: str
    @classmethod
    def from_days(cls, days: int) -> 'Tzolkin':
        t=days%260; number=(t%13)+1
        names=['Imix','Ik','Akbal','Kan','Chicchan','Cimi','Manik','Lamat',
               'Muluc','Oc','Chuen','Eb','Ben','Ix','Men','Cib','Caban',
               'Eznab','Cauac','Ahau']
        return cls(number, names[t%20])

@dataclass
class Haab:
    day: int; month: str
    MONTHS = ['Pop','Wo','Sip','Sotz','Sek','Xul',"Yaxk'in",'Mol',
              "Ch'en",'Yax','Sak','Keh','Mak',"K'ank'in",'Muwan',
              'Pax',"K'ayab","Kumk'u",'Wayeb']
    @classmethod
    def from_days(cls, days: int) -> 'Haab':
        h=days%365; return cls(h%20, cls.MONTHS[h//20])

# ============================================================
# π-GEOMETRY
# ============================================================

class PiGeometry:
    TAU = 2*math.pi

    @staticmethod
    def normalize_phase(phase: float) -> float:
        n=phase%PiGeometry.TAU; return n if n>=0 else n+PiGeometry.TAU

    @staticmethod
    def interference(p1,p2,a1=1.0,a2=1.0) -> float:
        diff=abs(p1-p2)%PiGeometry.TAU
        return math.cos(min(diff,PiGeometry.TAU-diff))*a1*a2

    @staticmethod
    def rotate_phase(phase,angle) -> float:
        return PiGeometry.normalize_phase(phase+angle)

# ============================================================
# TOKEN + FIELD
# ============================================================

@dataclass
class Token:
    position: np.ndarray; velocity: np.ndarray
    phase: float; phase_velocity: float
    long_count: LongCount; shard_id: int; coherence: float

    @classmethod
    def random(cls, dimensions=100) -> 'Token':
        return cls(
            position=np.random.randn(dimensions)*2,
            velocity=np.random.randn(dimensions)*0.1,
            phase=np.random.random()*2*math.pi,
            phase_velocity=np.random.random()*0.5,
            long_count=LongCount(
                np.random.randint(0,20),np.random.randint(0,20),
                np.random.randint(0,20),np.random.randint(0,18),
                np.random.randint(0,20)),
            shard_id=0, coherence=0.0)

@dataclass
class CardField:
    amplitude: np.ndarray; gradient: np.ndarray; curvature: np.ndarray
    pi_mod: float; maya_digit: int
    adjacency: List[int]; adjacency_strength: List[float]

# ============================================================
# π-KUHUL FIELD ENGINE
# ============================================================

class PiKuhulFieldEngine:
    """
    Field-theoretic RAG engine.
    Corpus = token states on 100D manifold.
    Retrieval = geodesic interference attention.
    Augment = shard transitions by phase coherence.
    Generate = field evolution step.
    """
    def __init__(self, num_tokens=1024, num_cards=52, dimensions=100,
                 dt=0.01, sigma=0.15, temperature=0.2):
        self.num_tokens=num_tokens; self.num_cards=num_cards
        self.dimensions=dimensions; self.dt=dt; self.sigma=sigma
        self.temperature=temperature
        self.tokens: List[Token]=[]
        self.cards: List[CardField]=[]
        self.pi_time=0.0
        self.global_long_count=LongCount(13,0,0,0,0)
        self.coherence_history=[]; self.shard_transitions=0
        self._initialize()

    def _initialize(self):
        for _ in range(self.num_tokens):
            t=Token.random(self.dimensions); t.shard_id=np.random.randint(0,self.num_cards)
            self.tokens.append(t)
        for i in range(self.num_cards):
            adj=[((i+j)%self.num_cards) for j in range(1,9)]
            self.cards.append(CardField(
                amplitude=np.random.randn(4)*2, gradient=np.random.randn(4)*0.5,
                curvature=np.random.randn(4)*0.2, pi_mod=np.random.random()*2,
                maya_digit=i%20, adjacency=adj,
                adjacency_strength=[1.0/j for j in range(1,9)]))

    def geodesic_distance(self,pos1,pos2,curvature) -> float:
        d=np.linalg.norm(pos1[:16]-pos2[:16])
        return d*(1+curvature[0]*d*d/6.0)

    def step(self) -> float:
        total_coherence=0.0
        self.pi_time=PiGeometry.normalize_phase(self.pi_time+self.dt*0.1)
        for i,token in enumerate(self.tokens):
            card=self.cards[token.shard_id%self.num_cards]
            token.phase=PiGeometry.rotate_phase(token.phase,self.dt*(card.pi_mod+self.pi_time*0.1))
            net_force=np.zeros(4); total_interference=0.0; shard_pot=0.0
            for j,other in enumerate(self.tokens):
                if i==j: continue
                other_card=self.cards[other.shard_id%self.num_cards]
                is_adj=(other.shard_id==token.shard_id or other.shard_id in card.adjacency)
                if not is_adj: continue
                dist=self.geodesic_distance(token.position[:16],other.position[:16],card.curvature)
                interf=PiGeometry.interference(token.phase,other.phase,
                    np.sum(card.amplitude),np.sum(other_card.amplitude))
                w=math.exp(-dist*dist/(2*self.sigma*self.sigma))*abs(interf)
                pd=other.position[:4]-token.position[:4]; n=np.linalg.norm(pd)
                if n>1e-6: net_force+=(pd/n)*w*(1.0 if interf>0 else -1.0)
                total_interference+=abs(interf)
                if interf>0.5 and other.shard_id!=token.shard_id: shard_pot+=interf
            if total_interference>0: net_force/=total_interference
            net_force+=card.gradient[:4]
            token.velocity[:4]=token.velocity[:4]*0.95+net_force*self.dt
            token.position[:4]+=token.velocity[:4]*self.dt
            pm=np.linalg.norm(token.position[:4])
            token.long_count=token.long_count.advance(max(1,int(abs(pm)*10)))
            if token.phase<self.dt*card.pi_mod or shard_pot>1.0:
                tgt=int(abs(token.position[0])*20)%20
                found=token.shard_id
                for adj in card.adjacency:
                    if adj<self.num_cards and self.cards[adj].maya_digit==tgt:
                        found=adj; break
                if np.random.random()<self.temperature:
                    found=card.adjacency[np.random.randint(0,min(8,len(card.adjacency)))]
                if found!=token.shard_id: self.shard_transitions+=1; token.shard_id=found
            token.coherence=token.coherence*0.99+(total_interference/100.0 if total_interference>0 else 0)
            total_coherence+=token.coherence
        self.global_long_count=self.global_long_count.advance(1)
        avg=total_coherence/self.num_tokens
        self.coherence_history.append(avg)
        return avg

    def run(self, steps=1000, callback=None) -> Dict:
        t0=datetime.now()
        for s in range(steps):
            c=self.step()
            if callback and s%10==0: callback(s,c,self.get_metrics())
            elif s%100==0: print(f"Step {s}: coherence={c*100:.1f}%")
        elapsed=(datetime.now()-t0).total_seconds()
        return {'final_coherence':self.coherence_history[-1] if self.coherence_history else 0,
                'avg_coherence':float(np.mean(self.coherence_history)) if self.coherence_history else 0,
                'shard_transitions':self.shard_transitions,
                'time_seconds':elapsed,'steps_per_second':steps/elapsed if elapsed>0 else 0}

    def get_metrics(self) -> Dict:
        lc=self.global_long_count
        return {
            'coherence':float(np.mean([t.coherence for t in self.tokens])),
            'avg_phase':float(np.mean([t.phase for t in self.tokens])),
            'pi_time':self.pi_time,'global_long_count':str(lc),
            'shard_transitions':self.shard_transitions,
            'active_tokens':sum(1 for t in self.tokens if t.coherence>0.1),
            'tzolkin':str(Tzolkin.from_days(lc.to_days())),
            'haab':str(Haab.from_days(lc.to_days()))
        }

    def to_optical_nodes(self):
        """
        Convert token states to optical node format for the SH wave lattice.
        Token.position[:3] → OpticalNode.pos
        Token.phase        → OpticalNode.sh[0] (band 0 phase)
        Token.coherence    → OpticalNode.energy()
        """
        nodes=[]
        for t in self.tokens:
            pos=t.position[:3]/np.linalg.norm(t.position[:3]+1e-9)
            sh=np.zeros(18)
            sh[0]=math.cos(t.phase)*t.coherence
            sh[1]=math.sin(t.phase)*t.coherence
            nodes.append({'pos':pos.tolist(),'sh':sh.tolist(),
                          'energy':t.coherence,'phase':t.phase,
                          'shard_id':t.shard_id,'long_count':str(t.long_count)})
        return nodes

if __name__=='__main__':
    import argparse
    p=argparse.ArgumentParser(description='pi-KUHUL Field RAG Engine')
    p.add_argument('--tokens',type=int,default=256)
    p.add_argument('--steps',type=int,default=500)
    p.add_argument('--temperature',type=float,default=0.2)
    args=p.parse_args()
    print(f"pi-KUHUL RAG Engine | {args.tokens} tokens | {args.steps} steps")
    engine=PiKuhulFieldEngine(num_tokens=args.tokens,temperature=args.temperature)
    results=engine.run(args.steps)
    m=engine.get_metrics()
    print(f"Coherence: {results['final_coherence']*100:.2f}%")
    print(f"Long Count: {m['global_long_count']}  Tzolk'in: {m['tzolkin']}  Haab: {m['haab']}")
    print(f"Shard transitions: {results['shard_transitions']}")

File: src/mupy/pi-kuhul/pi-kuhul-engine.js
// pi-kuhul-engine.js — π-KUHUL Field RAG Engine (ECMAScript port)
//
// THIS IS A RAG ENGINE where the retrieval space is curved:
//   Corpus    = Token states on 100D manifold (the "documents")
//   Retrieval = Geodesic phase interference (arccos, not cosine)
//   Augment   = Shard transitions by phase coherence (not doc stuffing)
//   Generate  = Field evolution step (wave propagation = generation)
//
// THE HAAB MONTHS = K'UHUL PHASES (not coincidence):
//   Haab months: Pop Wo Sip Sotz Sek Xul Yaxk'in Mol Ch'en Yax Sak Keh
//                Mak K'ank'in Muwan Pax K'ayab Kumk'u Wayeb
//   K'UHUL exec: Pop →      Wo →                Sek →     Ch'en → Xul
//   The calendar IS the execution schedule.
//
// 52 cards = SMGM-16 CardSlots = Calendar Round (52 Haab years = 18,980 days)
// Token.phase = OpticalNode.sh[] (SH wave state per band)
// shard_id = optical cluster membership (which icosphere region)
// interference() = our arccos(q·k^T) geodesic attention
// coherence_history = EntropyField values (high incoherence = high entropy)
//
// Connection to existing stack:
//   to_optical_nodes() → feed into OpticalProcessor / SVG3DComputeGraph
//   Token.long_count   → ARC replay tick counter
//   CardField.pi_mod   → pressure per fold (fold pressure mapper)
//   shard_transition   → ARC recorded between shards = ReplayableArc

const TAU = 2 * Math.PI;

// ─── Mayan digit system ───────────────────────────────────────────────────────

export const MAYAN_DIGITS = {
    '•':1,'••':2,'•••':3,'••••':4,
    '⎯':5,'⎯⎯':10,'⎯⎯⎯':15,
    '•⎯':6,'••⎯':7,'•••⎯':8,'••••⎯':9,
    '•⎯⎯':11,'••⎯⎯':12,'•••⎯⎯':13,'••••⎯⎯':14,
    '•⎯⎯⎯':16,'••⎯⎯⎯':17,'•••⎯⎯⎯':18,'••••⎯⎯⎯':19,
    '𑂐':0,'🐚':0,'○':0
};

const REVERSE_MAYAN = ['𑂐','•','••','•••','••••','⎯',
    '•⎯','••⎯','•••⎯','••••⎯','⎯⎯','•⎯⎯','••⎯⎯','•••⎯⎯','••••⎯⎯','⎯⎯⎯',
    '•⎯⎯⎯','••⎯⎯⎯','•••⎯⎯⎯','••••⎯⎯⎯'];

export function mayanToNumber(g) {
    if (MAYAN_DIGITS[g] !== undefined) return MAYAN_DIGITS[g];
    let v=0; for (const c of g) v=v*20+(MAYAN_DIGITS[c]??0); return v;
}
export function numberToMayan(n) {
    if (!Number.isSafeInteger(n)||n<0) return '𑂐';
    if (n===0) return '𑂐';
    const d=[]; let r=n;
    while(r>0){d.unshift(r%20);r=Math.floor(r/20);}
    return d.map(x=>REVERSE_MAYAN[x]).join('');
}

// ─── Maya calendar ────────────────────────────────────────────────────────────

// Haab months = K'UHUL phases. The calendar IS the execution schedule.
export const HAAB_MONTHS = [
    'Pop','Wo','Sip','Sotz','Sek','Xul',"Yaxk'in",'Mol',
    "Ch'en",'Yax','Sak','Keh','Mak',"K'ank'in",'Muwan',
    'Pax',"K'ayab","Kumk'u",'Wayeb'
];

export const MayanCalendar = {
    longCountToDays(b,k,t,u,kin){ return b*144000+k*7200+t*360+u*20+kin; },
    daysToLongCount(days){
        const b=Math.floor(days/144000); days%=144000;
        const k=Math.floor(days/7200);   days%=7200;
        const t=Math.floor(days/360);    days%=360;
        const u=Math.floor(days/20);     const kin=days%20;
        return {baktun:b,katun:k,tun:t,uinal:u,kin};
    },
    tzolkin(days){ return {number:(days%260%13)+1, dayIdx:days%260%20}; },
    haab(days){ return {day:days%365%20, month:HAAB_MONTHS[Math.floor(days%365/20)]}; },
    advance(lc, n=1){ return MayanCalendar.daysToLongCount(
        MayanCalendar.longCountToDays(lc.baktun,lc.katun,lc.tun,lc.uinal,lc.kin)+n); },
};

// ─── π-Geometry ───────────────────────────────────────────────────────────────

export const PiGeometry = {
    normalize(p){ const n=p%TAU; return n<0?n+TAU:n; },
    interference(p1,p2,a1=1,a2=1){
        const d=Math.abs(p1-p2)%TAU;
        return Math.cos(Math.min(d,TAU-d))*a1*a2;
    },
    rotate(phase,angle){ return PiGeometry.normalize(phase+angle); },
};

// ─── MayanMath ────────────────────────────────────────────────────────────────

export const MayanMath = {
    VIGESIMAL: 20,
    VENUS_SYNODIC: 584,
    CALENDAR_ROUND: 18980,
    ...MayanCalendar,
    normalizePhase: PiGeometry.normalize,
    normalizeTensor(t,scale=20){
        const n=Math.hypot(...t)||Number.EPSILON;
        return t.map(v=>v/n*scale);
    },
    coherence(phases,amps=null){
        let total=0; const n=phases.length;
        for(let i=0;i<n;i++) for(let j=i+1;j<n;j++){
            const d=Math.abs(phases[i]-phases[j])%TAU;
            total+=2*(amps?amps[i]*amps[j]:1)*Math.cos(Math.min(d,TAU-d));
        }
        return total/(n*(n-1)||1);
    },
};

// ─── PiKuhulFieldEngine ───────────────────────────────────────────────────────

export class PiKuhulFieldEngine {
    constructor({numTokens=256,numCards=52,dimensions=16,dt=0.01,sigma=0.15,temperature=0.2}={}) {
        this.numTokens=numTokens; this.numCards=numCards; this.dimensions=dimensions;
        this.dt=dt; this.sigma=sigma; this.temperature=temperature;
        this.tokens=[]; this.cards=[];
        this.piTime=0; this.tick=0;
        this.globalLC={baktun:13,katun:0,tun:0,uinal:0,kin:0};
        this.coherenceHistory=[]; this.shardTransitions=0;
        this._init();
    }

    _init(){
        for(let i=0;i<this.numTokens;i++){
            const pos=Array.from({length:this.dimensions},()=>(Math.random()-0.5)*4);
            const vel=Array.from({length:this.dimensions},()=>(Math.random()-0.5)*0.2);
            this.tokens.push({pos,vel,phase:Math.random()*TAU,coherence:0,
                shardId:Math.floor(Math.random()*this.numCards),
                lc:{baktun:Math.floor(Math.random()*20),katun:Math.floor(Math.random()*20),
                    tun:Math.floor(Math.random()*20),uinal:Math.floor(Math.random()*18),
                    kin:Math.floor(Math.random()*20)}});
        }
        for(let i=0;i<this.numCards;i++){
            const adj=Array.from({length:8},(_,j)=>(i+j+1)%this.numCards);
            this.cards.push({amplitude:Array.from({length:4},()=>Math.random()*2-1),
                gradient:Array.from({length:4},()=>Math.random()*0.5-0.25),
                piMod:Math.random()*2, mayaDigit:i%20, adjacency:adj});
        }
    }

    _geodesicDist(a,b){
        let s=0; const n=Math.min(a.length,b.length,4);
        for(let i=0;i<n;i++){const d=(a[i]-b[i])/20;s+=d*d;}
        const flat=Math.sqrt(s);
        return flat*(1+0.1*flat*flat/6);
    }

    step(){
        this.piTime=PiGeometry.normalize(this.piTime+this.dt*0.1);
        let totalCoh=0;
        for(let i=0;i<this.tokens.length;i++){
            const tok=this.tokens[i];
            const card=this.cards[tok.shardId%this.numCards];
            tok.phase=PiGeometry.rotate(tok.phase,this.dt*(card.piMod+this.piTime*0.1));
            const force=[0,0,0,0]; let totalInterf=0, shardPot=0;
            for(let j=0;j<this.tokens.length;j++){
                if(i===j) continue;
                const other=this.tokens[j];
                const otherCard=this.cards[other.shardId%this.numCards];
                const isAdj=other.shardId===tok.shardId||card.adjacency.includes(other.shardId);
                if(!isAdj) continue;
                const dist=this._geodesicDist(tok.pos,other.pos);
                const interf=PiGeometry.interference(tok.phase,other.phase,
                    card.amplitude.reduce((s,v)=>s+v,0)/4,
                    otherCard.amplitude.reduce((s,v)=>s+v,0)/4);
                const w=Math.exp(-dist*dist/(2*this.sigma*this.sigma))*Math.abs(interf);
                const d=other.pos.slice(0,4).map((v,k)=>v-tok.pos[k]);
                const dn=Math.hypot(...d)||1e-9;
                const sign=interf>0?1:-1;
                for(let k=0;k<4;k++) force[k]+=(d[k]/dn)*w*sign;
                totalInterf+=Math.abs(interf);
                if(interf>0.5&&other.shardId!==tok.shardId) shardPot+=interf;
            }
            if(totalInterf>0) for(let k=0;k<4;k++) force[k]/=totalInterf;
            for(let k=0;k<4;k++) force[k]+=card.gradient[k]||0;
            for(let k=0;k<4;k++){
                tok.vel[k]=tok.vel[k]*0.95+force[k]*this.dt;
                tok.pos[k]+=tok.vel[k]*this.dt;
            }
            const pm=Math.hypot(...tok.pos.slice(0,4));
            tok.lc=MayanCalendar.advance(tok.lc,Math.max(1,Math.floor(Math.abs(pm)*10)));
            if(tok.phase<this.dt*card.piMod||shardPot>1.0){
                const tgt=Math.floor(Math.abs(tok.pos[0])*20)%20;
                let found=tok.shardId;
                for(const adj of card.adjacency){
                    if(adj<this.numCards&&this.cards[adj].mayaDigit===tgt){found=adj;break;}
                }
                if(Math.random()<this.temperature)
                    found=card.adjacency[Math.floor(Math.random()*Math.min(8,card.adjacency.length))];
                if(found!==tok.shardId){this.shardTransitions++;tok.shardId=found;}
            }
            tok.coherence=tok.coherence*0.99+(totalInterf>0?totalInterf/100:0);
            totalCoh+=tok.coherence;
        }
        this.globalLC=MayanCalendar.advance(this.globalLC);
        this.tick++;
        const avg=totalCoh/this.numTokens;
        this.coherenceHistory.push(avg);
        return avg;
    }

    run(steps=500){
        for(let s=0;s<steps;s++) this.step();
        const avg=this.coherenceHistory.reduce((a,b)=>a+b,0)/this.coherenceHistory.length;
        return {finalCoherence:this.coherenceHistory.at(-1)??0,avgCoherence:avg,
                shardTransitions:this.shardTransitions,ticks:this.tick};
    }

    // Convert to optical node format for SVG3DComputeGraph.fromOpticalMesh()
    toOpticalNodes(){
        return this.tokens.map(t=>{
            const n=Math.hypot(...t.pos.slice(0,3))||1;
            const sh=new Float32Array(18);
            sh[0]=Math.cos(t.phase)*t.coherence;
            sh[1]=Math.sin(t.phase)*t.coherence;
            return {position:[t.pos[0]/n,t.pos[1]/n,t.pos[2]/n],
                    normal:[t.pos[0]/n,t.pos[1]/n,t.pos[2]/n],
                    uv:[0,0], sh, energy:t.coherence};
        });
    }

    metrics(){
        const lc=this.globalLC;
        const days=MayanCalendar.longCountToDays(lc.baktun,lc.katun,lc.tun,lc.uinal,lc.kin);
        const tz=MayanCalendar.tzolkin(days), haab=MayanCalendar.haab(days);
        const avg=this.tokens.reduce((s,t)=>s+t.coherence,0)/this.numTokens;
        return {coherence:avg,piTime:this.piTime,tick:this.tick,
                longCount:`${lc.baktun}.${lc.katun}.${lc.tun}.${lc.uinal}.${lc.kin}`,
                tzolkin:`${tz.number} ${['Imix','Ik','Akbal','Kan','Chicchan','Cimi','Manik','Lamat','Muluc','Oc','Chuen','Eb','Ben','Ix','Men','Cib','Caban','Eznab','Cauac','Ahau'][tz.dayIdx]}`,
                haab:`${haab.day} ${haab.month}`,
                shardTransitions:this.shardTransitions,
                activeTokens:this.tokens.filter(t=>t.coherence>0.1).length};
    }
}
